### DISCUSSION TITLE
Using Kawaii Physics with the Epic Games Animation Sample Motion Matching

### DISCUSSION THREAD

**Discussion by TimothyJR** [pafuhana1213/KawaiiPhysics#136] [Q&A] [answered]
I set up Kawaii Physics with the animation sample project for a long braid. The project has two forms of movement. 

One where the character always faces the direction the camera faces and one where the character faces the direction they are moving. Everything works fine for the one where the character faces the direction of the camera. 

But when using the movement where the character faces the direction of movement, whenever I have the character turn around (180 degree turn) the kawaii physics bones will stutter and flip 180 degrees for one frame before returning to where they should be. Below I attached a video of what I am talking about. You can see when each turn starts, the hair jitters to the other side of the character.

https://github.com/user-attachments/assets/3f9de682-1e0b-495a-8ce3-7ba2fa7852c9

**Comment by pafuhana1213** [comment]
Hi @TimothyJR 

I tried to repro you issue with echo, but unfortunately failed it 
https://github.com/user-attachments/assets/2cd9b674-5ddf-465b-a5cc-3f7b042a6a15

Please provide any information I or others need to diagnose your problem, such as your model/bone configuration, any changes you've made from Kawaii's default settings, etc.

Also, if you haven't edited Kawaii's teleport settings yet, that might help you.

thanks

**Reply by TimothyJR** [reply]
Hello, thanks for the quick reply. I started a new project with just the Game Animation Motion Matching sample. I used Echo to make it as similar to yours as possible. From the blank project, I opened Echo's animation blueprint. I removed the rigidbody node and put a Kawaii Physics node in its place. I assigned C_Ponytail_B_Root_Jnt as the Root bone. Then I recorded this video. While it isn't as obvious as the one with my character, you can still see bones snapping when starting to turn. I have tried different settings, but nothing seems to stop this from happening. I added these settings one at a time starting from default:

Damping -> .35
Stiffness -> 0.1
World Damping Location & Rotation -> .9 -> 1
Dummy Bone Length -> 15
Limit Angle -> 60
Teleport Distance Threshold -> 0
Teleport Rotation Threshold -> 0


https://github.com/user-attachments/assets/b5a6cf35-e0eb-4364-920b-af05c48493df

**Reply by reyHansss** [reply]
I ran into the same problem, it seems that the character's turn was set to immediately face the desired position while on the ground, Resulting in significant changes in the bones.
The capsule of the character rotates, but the actual model of the character does not rotate immediately due to the root bone shift. But the dynamic bone was affected by the rotation of the capsule body.
You can find the relevant note in the character blueprint: When the character is on the ground, this function sets the Rotation Rate to -1, which causes the character to rotate instantly.
You can fix this by changing -1 to something else.But that doesn't work very well. Waiting for the next update of epic game.
Can you give a switch to control whether the dynamic bone is affected by rotation？

**Reply by TimothyJR** [reply]
Ah thank you! You are absolutely correct in that the instant rotation is causing the jitter effect. I'm not sure on what I will settle on for a solution to this yet, since I only had enough time to confirm that it was the cause. But this is a great start. Thanks again.

**Reply by pafuhana1213** [reply]
Thanks to gread guys! I now understand the problem.

KawaiiPhysics processes in ComponentSpace, but the root of the ComponentSpace moves a lot during the turn, which is why this problem occurs. Therefore, the same problem occurs with a RigidBody node where SimulationSpace is changed to Component Space.

To solve this correctly, KawaiiPhysics needs to support WorldSpace and Base Bone Space. However, this would be a bit of a big change, so I'll start by looking into a simple solution. Please wait a bit.
https://github.com/user-attachments/assets/d07e659d-addb-4f86-b553-49e4cb27d1c8

**Reply by yixingming** [reply]
I have a result for this problem? are you interested ？

**Reply by yixingming** [reply]
yixingming:newRequest

**Comment by LoBDo** [comment]
Hello! @pafuhana1213 
First of all, thank you so much for your continuous updates.
Could you let me know if this issue has been resolved?

**Comment by pafuhana1213** [comment, marked as answer]
Hi guys,

Sorry for my late support... I've completed to support World/BaseBoneSpace to fix this issue!
Please check it out!
https://github.com/pafuhana1213/KawaiiPhysics/discussions/170

thanks!
